platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/mappingToJvm/RuntimeFunctionResolver.kt

```diff
@@ -44,44 +47,33 @@ class MemberFunctionResolver(private val configureLambdaHandler: ConfigureLambda
         val parameterBindingStub = schemaFunction.parameters.associateWith { Any() }
         val hasConfigureLambda = (schemaFunction.semantics as? FunctionSemantics.ConfigureSemantics)?.configureBlockRequirement?.allows ?: false
 
+        fun signature(function: KFunction<*>): String =
+            function::class.memberProperties.first { it.name == "signature" }.apply { isAccessible = true }.call(function) as String
+
+        fun matchesAnnotationsRecursively(function: KFunction<*>, receiverClass: KClass<*>, predicate: (Annotation) -> Boolean): Boolean =
+            if (function.annotations.any(predicate)) {
+                true
+            } else {
+                val functionSignature = signature(function)
+                val parentClasses = receiverClass.supertypes.mapNotNull { it.classifier }.filterIsInstance<KClass<*>>()
+                parentClasses.any { parentClass ->
+                    val parentFunction = parentClass.memberFunctions.firstOrNull { signature(it) == functionSignature }
+                    parentFunction?.let { matchesAnnotationsRecursively(it, parentClass, predicate) } ?: false
+                }
+            }
+
         val resolutions = receiverClass.memberFunctions
             .filter { function -> function.name == schemaFunction.simpleName && FunctionBinding.convertBinding(function, Any(), parameterBindingStub, hasConfigureLambda, configureLambdaHandler) != null }
+            .filter { f -> matchesAnnotationsRecursively(f, receiverClass) { it is Builder || it is Configuring || it is Adding || it is Restricted || it is HasDefaultValue } }
             .toList()
 
         return when {
             resolutions.isEmpty() -> RuntimeFunctionResolver.Resolution.Unresolved
             resolutions.size == 1 -> RuntimeFunctionResolver.Resolution.Resolved(ReflectionFunction(resolutions[0], configureLambdaHandler))
-            else -> {
-                val refinedResolutions = resolutions.filter { function -> parametersMatch(function, schemaFunction) }
-                val finalResolution = if (refinedResolutions.size == 1) refinedResolutions[0] else
-                    error("Failed disambiguating between following functions: ${resolutions.joinToString(prefix = "\n\t", separator = "\n\t") { f -> f.toString() }}")
-                return RuntimeFunctionResolver.Resolution.Resolved(ReflectionFunction(finalResolution, configureLambdaHandler))
-            }
+            else -> error("Failed disambiguating between following functions: ${resolutions.joinToString(prefix = "\n\t", separator = "\n\t") { f -> f.toString() }}")
         }
 
     }
-
-    @OptIn(ExperimentalStdlibApi::class)
-    private fun parametersMatch(function: KFunction<*>, schemaFunction: SchemaFunction): Boolean {
```

Comment:
💭 To me, it seems that the annotation-based approach should complement the type-based one rather than replace it.

There are still potential cases in which a user would declare and annotate more than one function overload, and their signatures would differ only in the parameter types, e.g.:

```kotlin
@Restricted
fun f(data: Int): MyData = ...

@Restricted
fun f(data: String): MyData = ...
```

In this case, the DCL function call resolver will be able to resolve the right overload, and so the runtime function resolver should be able to choose the right function, too.